### ISSUE TITLE
TImers Alarms not sounding

### ISSUE THREAD

**Issue by mw1fly** [alexa/avs-device-sdk#1183]
## Briefly summarize your issue:

A timer or alarm does not sound when it should,  sometimes it will sound half an hour or so later.

## What is the expected behavior?

Timer or alarm should sound when ended.

## What behavior are you observing?

The alarm or timer does not sound or sometimes sounds up to an hour later.

## Provide the steps to reproduce the issue, if applicable:

Set a timer and wait.

### What version of the AVS Device SDK are you using?
      v1.11.0 released 12/19/2018

I using a raspberry pi,  I have installed AVS via the install scripts setup.sh, genConfig.sh and pi.sh as per the guide.

When running the test.sh script it fails on the following errors.
298:TimeTest.testStringConversion
301:TimeTest.testTimeConversion

298/1168 Test: TimeTest.testStringConversion
Command: "/home/rich/build/AVSCommon/Utils/test/TimeUtilsTest" "--gtest_filter=TimeTest.testStringConversion"
Directory: /home/rich/build/AVSCommon/Utils/test
"TimeTest.testStringConversion" start time: Jan 22 01:59 GMT
Output:
----------------------------------------------------------
2019-01-22 01:59:09.313 [  1] I sdkVersion: 1.11.0
Running main() from gmock_main.cc
Note: Google Test filter = TimeTest.testStringConversion
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TimeTest
[ RUN      ] TimeTest.testStringConversion
/home/rich/avs-device-sdk/AVSCommon/Utils/test/TimeUtilsTest.cpp:44: Failure
      Expected: dateTm.tm_hour
      Which is: 22
To be equal to: 21
[  FAILED  ] TimeTest.testStringConversion (1 ms)
[----------] 1 test from TimeTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TimeTest.testStringConversion

 1 FAILED TEST
-------------------------------------------------
<end of output>
Test time =   0.12 sec
----------------------------------------------------------
Test Failed.
"TimeTest.testStringConversion" end time: Jan 22 01:59 GMT
"TimeTest.testStringConversion" time elapsed: 00:00:00
----------------------------------------------------------
301/1168 Test: TimeTest.testTimeConversion
Command: "/home/rich/build/AVSCommon/Utils/test/TimeUtilsTest" "--gtest_filter=TimeTest.testTimeConversion"
Directory: /home/rich/build/AVSCommon/Utils/test
"TimeTest.testTimeConversion" start time: Jan 22 01:59 GMT
Output:
----------------------------------------------------------
2019-01-22 01:59:09.560 [  1] I sdkVersion: 1.11.0
Running main() from gmock_main.cc
Note: Google Test filter = TimeTest.testTimeConversion
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TimeTest
[ RUN      ] TimeTest.testTimeConversion
/home/rich/avs-device-sdk/AVSCommon/Utils/test/TimeUtilsTest.cpp:73: Failure
      Expected: randomDate
      Which is: 524089800
To be equal to: convertBack
      Which is: 524093400
[  FAILED  ] TimeTest.testTimeConversion (0 ms)
[----------] 1 test from TimeTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TimeTest.testTimeConversion

 1 FAILED TEST
<end of output>
Test time =   0.07 sec
----------------------------------------------------------
Test Failed.
"TimeTest.testTimeConversion" end time: Jan 22 01:59 GMT
"TimeTest.testTimeConversion" time elapsed: 00:00:00
----------------------------------------------------------

### Tell us what hardware you're using:
- [ ] Desktop / Laptop
- [x] Raspberry Pi
- [ ] Other - tell us more:

### Tell us about your OS (Type & version):
- [ ] Linux
- [ ] MacOS 
- [x] Raspbian Stretch
- [ ] Raspbian Jessy
- [ ] Other - tell us more:

**Comment by visayamv** [comment]
Hi @mw1fly , to help us debug, can you please run the unit test through gdb?

1. install gdb
sudo apt-get install gdb

2. Go to the build directory (you can check this in the test.sh script)
example: cd /home/pi/user/build

3. Execute: ctest -VV -R TimeTest.testStringConversion$

There should appear something like this:
`
285: Test command: /home/pi/user/build/AVSCommon/Utils/test/TimeUtilsTest "--gtest_filter=TimeTest.testStringConversion"
`

Make note of the test command. In this case it's: 

`/home/pi/user/build/AVSCommon/Utils/test/TimeUtilsTest "--gtest_filter=TimeTest.testStringConversion"`

4. Run the test command through gdb:

`gdb --args <test command>`

example:

`gdb --args /home/pi/user/build/AVSCommon/Utils/test/TimeUtilsTest "--gtest_filter=TimeTest.testStringConversion"`

5. You will go to the (gdb) prompt.

6. Set breakpoints:

b TimeUtils.cpp:135
b TimeUtils.cpp:203

7. Run the test

run

8. When it stops at breakpoint 1 (TimeUtils.cpp:135), do:

info args
print(converted)
print(offset)
continue

9. It should stop on the second breakpoint (TimeUtils.cpp:203), now do:

info args
print(convertedTimeT)
quit

And post the results here. Thanks!

**Comment by mw1fly** [comment]
Hi Visayanv,  here are the results I got as per your instructions.

rich@raspberrypi:~/build $ gdb --args /home/rich/build/AVSCommon/Utils/test/TimeUtilsTest "--gtest_filter=TimeTest.testStringConversion"
GNU gdb (Raspbian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "arm-linux-gnueabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/rich/build/AVSCommon/Utils/test/TimeUtilsTest...done.
(gdb) b TimeUtils.cpp:135
No source file named TimeUtils.cpp.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (TimeUtils.cpp:135) pending.
(gdb) b TimeUtils.cpp:203
No source file named TimeUtils.cpp.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (TimeUtils.cpp:203) pending.
(gdb) run
Starting program: /home/rich/build/AVSCommon/Utils/test/TimeUtilsTest --gtest_filter=TimeTest.testStringConversion
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
2019-01-25 08:55:30.608 [  1] I sdkVersion: 1.11.0
Running main() from gmock_main.cc
Note: Google Test filter = TimeTest.testStringConversion
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TimeTest
[ RUN      ] TimeTest.testStringConversion

Breakpoint 1, alexaClientSDK::avsCommon::utils::timing::TimeUtils::convertToUtcTimeT (this=0x7effeedc, utcTm=0x7effe2ec, ret=0x7effe2e8)
    at /home/rich/avs-device-sdk/AVSCommon/Utils/src/TimeUtils.cpp:135
135	    return true;
(gdb) info args
this = 0x7effeedc
utcTm = 0x7effe2ec
ret = 0x7effe2e8
(gdb) print(converted)
$1 = 524093400
(gdb) print(offset)
$2 = -3600
(gdb) continue
Continuing.

Breakpoint 2, alexaClientSDK::avsCommon::utils::timing::TimeUtils::convert8601TimeStringToUnix (this=0x7effeedc, timeString="1986-08-10T21:30:00+0000", 
    convertedTime=0x7effeeb8) at /home/rich/avs-device-sdk/AVSCommon/Utils/src/TimeUtils.cpp:203
203	    if (!ok) {
(gdb) info args
this = 0x7effeedc
timeString = "1986-08-10T21:30:00+0000"
convertedTime = 0x7effeeb8
(gdb) print(convertedTimeT)
$3 = 524097000
(gdb) quit
A debugging session is active.

	Inferior 1 [process 14405] will be killed.

Quit anyway? (y or n)

**Comment by mw1fly** [comment]
Hi,
Its Europe/London

Regards
Richard.

**Comment by womw** [comment]
Hey @mw1fly thanks for keeping us updated.

We are investigating this internally and will update once we have more information.

**Comment by mavamazon** [comment]
Hi @mw1fly!

Thank you for reporting this! We have found the root cause. The fix is going to be published in one of the future versions.

**Comment by kclchan** [comment]
Thanks for opening this issue.  It has been fixed in our v1.12 release.  Thanks!

**Comment by xcraftster** [comment]
Same problem here guys. I am on **Raspbian Buster Lite** (Linux raspberrypi 4.19.108-v7+ #1298 SMP Fri Mar 6 18:04:35 GMT 2020 armv7l GNU/Linux). AVS SDK - **1.18.0**

The date and timezone on Raspberry is ok. I checked it with date command. When you ask Alexa about he time and date it says that correctly.

So, no alarms, timers nor reminders working in my case. 

There is another glitch - I set up a reminder in Alexa app on Android, the reminder did not work. I then rebooted Pi with Alexa and it announced this reminded right after booting. Then I deleted this reminder in Alexa app and rebooted my Pi with Alexa and it announced this reminder again!
